File: libs/go/scanning.go
package spade

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"

	"gopkg.in/yaml.v3"
)

// InputKind distinguishes single-file from multi-file input entries.
type InputKind int

const (
	InputSingle   InputKind = iota
	InputMultiple
)

// InputEntry represents a scanned input from the inputs/ directory.
type InputEntry struct {
	Kind  InputKind
	Path  string   // populated for InputSingle
	Paths []string // populated for InputMultiple
}

// LoadParams reads and parses params.yaml from the given base directory.
// Returns an empty map if the file does not exist or is empty.
func LoadParams(base string) (map[string]any, error) {
	path := filepath.Join(base, "params.yaml")
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return map[string]any{}, nil
		}
		return nil, err
	}
	var params map[string]any
	if err := yaml.Unmarshal(data, &params); err != nil {
		return nil, err
	}
	if params == nil {
		return map[string]any{}, nil
	}
	return params, nil
}

// ScanInputs reads the inputs/ directory at the given base path and returns
// a map of input name to InputEntry.
func ScanInputs(base string) (map[string]InputEntry, error) {
	inputsDir := filepath.Join(base, "inputs")
	subdirs, err := os.ReadDir(inputsDir)
	if err != nil {
		if os.IsNotExist(err) {
			return map[string]InputEntry{}, nil
		}
		return nil, err
	}

	// Sort subdirectories by name
	sort.Slice(subdirs, func(i, j int) bool {
		return subdirs[i].Name() < subdirs[j].Name()
	})

	result := make(map[string]InputEntry)
	for _, subdir := range subdirs {
		if !subdir.IsDir() {
			continue
		}
		name := subdir.Name()
		subdirPath := filepath.Join(inputsDir, name)

		entries, err := os.ReadDir(subdirPath)
		if err != nil {
			return nil, err
		}

		var files []string
		for _, entry := range entries {
			if !entry.IsDir() {
				files = append(files, filepath.Join(subdirPath, entry.Name()))
			}
		}
		sort.Strings(files)

		if len(files) == 0 {
			return nil, &ErrEmptyInputDir{Name: name}
		}

		if len(files) == 1 {
			result[name] = InputEntry{Kind: InputSingle, Path: files[0]}
		} else {
			result[name] = InputEntry{Kind: InputMultiple, Paths: files}
		}
	}
	return result, nil
}

// Args holds merged parameters and inputs for handler consumption.
type Args struct {
	params map[string]any
	inputs map[string]InputEntry
}

// HasInput checks whether an input with the given name exists.
func (a *Args) HasInput(name string) bool {
	_, ok := a.inputs[name]
	return ok
}

// HasParam checks whether a parameter with the given name exists.
func (a *Args) HasParam(name string) bool {
	_, ok := a.params[name]
	return ok
}

// BuildArgs constructs an Args struct from the given base path by loading
// params.yaml and scanning the inputs/ directory.
func BuildArgs(base string) (*Args, error) {
	params, err := LoadParams(base)
	if err != nil {
		return nil, err
	}
	inputs, err := ScanInputs(base)
	if err != nil {
		return nil, err
	}
	return &Args{params: params, inputs: inputs}, nil
}

// Input retrieves a typed file/directory input by name from Args.
// T must be a pointer to a type implementing FromInput.
func Input[T interface {
	*E
	FromInput
}, E any](args *Args, name string) (E, error) {
	var zero E
	entry, ok := args.inputs[name]
	if !ok {
		return zero, &ErrInputNotFound{Name: name}
	}
	val := T(&zero)
	switch entry.Kind {
	case InputSingle:
		if err := val.FromSingleFile(entry.Path); err != nil {
			return zero, err
		}
	case InputMultiple:
		if err := val.FromMultipleFiles(entry.Paths); err != nil {
			return zero, err
		}
	}
	return zero, nil
}

// Param retrieves a typed scalar parameter by name from Args.
func Param[T any](args *Args, name string) (T, error) {
	var zero T
	raw, ok := args.params[name]
	if !ok {
		return zero, &ErrParamNotFound{Name: name}
	}

	// Try direct type assertion first
	if v, ok := raw.(T); ok {
		return v, nil
	}

	// Handle numeric conversions: yaml.v3 parses integers as int, floats as float64
	switch any(zero).(type) {
	case float64:
		switch v := raw.(type) {
		case int:
			return any(float64(v)).(T), nil
		case int64:
			return any(float64(v)).(T), nil
		}
	case int:
		switch v := raw.(type) {
		case float64:
			return any(int(v)).(T), nil
		case int64:
			return any(int(v)).(T), nil
		}
	case int64:
		switch v := raw.(type) {
		case int:
			return any(int64(v)).(T), nil
		case float64:
			return any(int64(v)).(T), nil
		}
	}

	// Fallback: marshal to YAML then unmarshal into T
	data, err := yaml.Marshal(raw)
	if err != nil {
		return zero, fmt.Errorf("parameter '%s': cannot convert value: %w", name, err)
	}
	var result T
	if err := yaml.Unmarshal(data, &result); err != nil {
		return zero, fmt.Errorf("parameter '%s': cannot convert to target type: %w", name, err)
	}
	return result, nil
}

File: libs/rust/src/scanning.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

use crate::error::{Result, SpadeError};
use crate::types::FromInput;

/// Raw representation of a scanned input entry from the `inputs/` directory.
#[derive(Debug, Clone)]
pub enum InputEntry {
    /// A single file was found in the input subdirectory.
    SingleFile(String),
    /// Multiple files were found in the input subdirectory.
    MultipleFiles(Vec<String>),
}

/// Load scalar parameters from `params.yaml` at the given base path.
pub fn load_params_from(base: &Path) -> Result<HashMap<String, serde_yaml::Value>> {
    let params_path = base.join("params.yaml");
    if !params_path.exists() {
        return Ok(HashMap::new());
    }
    let contents = fs::read_to_string(params_path)?;
    let value: serde_yaml::Value = serde_yaml::from_str(&contents)?;
    match value {
        serde_yaml::Value::Mapping(map) => {
            let mut result = HashMap::new();
            for (k, v) in map {
                if let serde_yaml::Value::String(key) = k {
                    result.insert(key, v);
                }
            }
            Ok(result)
        }
        serde_yaml::Value::Null => Ok(HashMap::new()),
        _ => Ok(HashMap::new()),
    }
}

/// Load scalar parameters from `params.yaml` in the current working directory.
pub fn load_params() -> Result<HashMap<String, serde_yaml::Value>> {
    load_params_from(Path::new("."))
}

/// Scan the `inputs/` directory at the given base path.
pub fn scan_inputs_from(base: &Path) -> Result<HashMap<String, InputEntry>> {
    let inputs_dir = base.join("inputs");
    if !inputs_dir.exists() {
        return Ok(HashMap::new());
    }

    let mut entries: Vec<_> = fs::read_dir(&inputs_dir)?
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().is_ok_and(|ft| ft.is_dir()))
        .collect();
    entries.sort_by_key(|e| e.file_name());

    let mut result = HashMap::new();
    for entry in entries {
        let name = entry.file_name().to_string_lossy().to_string();
        let subdir = entry.path();

        let mut files: Vec<String> = fs::read_dir(&subdir)?
            .filter_map(|e| e.ok())
            .filter(|e| e.file_type().is_ok_and(|ft| ft.is_file()))
            .map(|e| e.path().to_string_lossy().to_string())
            .collect();
        files.sort();

        if files.is_empty() {
            return Err(SpadeError::EmptyInputDir { name });
        }

        let input_entry = if files.len() == 1 {
            InputEntry::SingleFile(files.into_iter().next().unwrap())
        } else {
            InputEntry::MultipleFiles(files)
        };

        result.insert(name, input_entry);
    }

    Ok(result)
}

/// Scan the `inputs/` directory in the current working directory.
pub fn scan_inputs() -> Result<HashMap<String, InputEntry>> {
    scan_inputs_from(Path::new("."))
}

/// Merged arguments from `params.yaml` and `inputs/`, ready for handler consumption.
pub struct Args {
    params: HashMap<String, serde_yaml::Value>,
    inputs: HashMap<String, InputEntry>,
}

impl Args {
    /// Retrieve a typed file/directory input by name.
    pub fn input<T: FromInput>(&self, name: &str) -> Result<T> {
        let entry = self
            .inputs
            .get(name)
            .ok_or_else(|| SpadeError::InputNotFound {
                name: name.to_string(),
            })?;

        match entry {
            InputEntry::SingleFile(path) => T::from_single_file(path.clone()),
            InputEntry::MultipleFiles(paths) => T::from_multiple_files(paths.clone()),
        }
    }

    /// Retrieve a typed scalar parameter by name.
    pub fn param<T: serde::de::DeserializeOwned>(&self, name: &str) -> Result<T> {
        let value = self
            .params
            .get(name)
            .ok_or_else(|| SpadeError::ParamNotFound {
                name: name.to_string(),
            })?;

        serde_yaml::from_value(value.clone()).map_err(SpadeError::YamlError)
    }

    /// Check whether an input with the given name exists.
    pub fn has_input(&self, name: &str) -> bool {
        self.inputs.contains_key(name)
    }

    /// Check whether a parameter with the given name exists.
    pub fn has_param(&self, name: &str) -> bool {
        self.params.contains_key(name)
    }
}

/// Build the `Args` struct from the given base path.
pub fn build_args_from(base: &Path) -> Result<Args> {
    let params = load_params_from(base)?;
    let inputs = scan_inputs_from(base)?;
    Ok(Args { params, inputs })
}

/// Build the `Args` struct from the current working directory.
pub fn build_args() -> Result<Args> {
    build_args_from(Path::new("."))
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{Directory, File, RasterFile, RasterFileCollection};
    use tempfile::TempDir;

    fn setup_work_dir() -> TempDir {
        let dir = TempDir::new().unwrap();
        fs::create_dir_all(dir.path().join("inputs")).unwrap();
        fs::create_dir_all(dir.path().join("outputs")).unwrap();
        fs::create_dir_all(dir.path().join("logs")).unwrap();
        dir
    }

    fn write_params(dir: &TempDir, content: &str) {
        fs::write(dir.path().join("params.yaml"), content).unwrap();
    }

    fn create_input_file(dir: &TempDir, name: &str, filename: &str, content: &[u8]) {
        let input_dir = dir.path().join("inputs").join(name);
        fs::create_dir_all(&input_dir).unwrap();
        fs::write(input_dir.join(filename), content).unwrap();
    }

    // -- load_params tests --

    #[test]
    fn test_load_params_basic() {
        let dir = setup_work_dir();
        write_params(&dir, "buffer_distance: 30\nmethod: bilinear\n");
        let params = load_params_from(dir.path()).unwrap();
        assert_eq!(
            params["buffer_distance"],
            serde_yaml::Value::Number(30.into())
        );
        assert_eq!(
            params["method"],
            serde_yaml::Value::String("bilinear".into())
        );
    }

    #[test]
    fn test_load_params_empty_file() {
        let dir = setup_work_dir();
        write_params(&dir, "");
        let params = load_params_from(dir.path()).unwrap();
        assert!(params.is_empty());
    }

    #[test]
    fn test_load_params_missing_file() {
        let dir = setup_work_dir();
        let params = load_params_from(dir.path()).unwrap();
        assert!(params.is_empty());
    }

    // -- scan_inputs tests --

    #[test]
    fn test_scan_inputs_single_file() {
        let dir = setup_work_dir();
        create_input_file(&dir, "raster", "data.tif", b"test data");
        let inputs = scan_inputs_from(dir.path()).unwrap();
        assert!(matches!(inputs["raster"], InputEntry::SingleFile(_)));
        if let InputEntry::SingleFile(path) = &inputs["raster"] {
            assert!(path.ends_with("data.tif"));
        }
    }

    #[test]
    fn test_scan_inputs_multiple_files() {
        let dir = setup_work_dir();
        create_input_file(&dir, "tiles", "001.tif", b"data");
        create_input_file(&dir, "tiles", "002.tif", b"data");
        create_input_file(&dir, "tiles", "003.tif", b"data");
        let inputs = scan_inputs_from(dir.path()).unwrap();
        assert!(matches!(inputs["tiles"], InputEntry::MultipleFiles(_)));
        if let InputEntry::MultipleFiles(paths) = &inputs["tiles"] {
            assert_eq!(paths.len(), 3);
        }
    }

    #[test]
    fn test_scan_inputs_empty_dir_error() {
        let dir = setup_work_dir();
        fs::create_dir_all(dir.path().join("inputs/empty")).unwrap();
        let result = scan_inputs_from(dir.path());
        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(matches!(err, SpadeError::EmptyInputDir { .. }));
    }

    #[test]
    fn test_scan_inputs_no_inputs_dir() {
        let dir = TempDir::new().unwrap();
        // No inputs/ directory
        let inputs = scan_inputs_from(dir.path()).unwrap();
        assert!(inputs.is_empty());
    }

    #[test]
    fn test_scan_inputs_multiple_subdirs() {
        let dir = setup_work_dir();
        create_input_file(&dir, "reference", "ref.tif", b"ref");
        create_input_file(&dir, "target", "tgt.tif", b"tgt");
        let inputs = scan_inputs_from(dir.path()).unwrap();
        assert!(inputs.contains_key("reference"));
        assert!(inputs.contains_key("target"));
    }

    // -- Args tests --

    #[test]
    fn test_args_input_raster_file() {
        let dir = setup_work_dir();
        create_input_file(&dir, "source", "data.tif", b"test");
        let args = build_args_from(dir.path()).unwrap();
        let raster: RasterFile = args.input("source").unwrap();
        assert!(raster.path.ends_with("data.tif"));
    }

    #[test]
    fn test_args_input_collection() {
        let dir = setup_work_dir();
        create_input_file(&dir, "tiles", "a.tif", b"a");
        create_input_file(&dir, "tiles", "b.tif", b"b");
        let args = build_args_from(dir.path()).unwrap();
        let coll: RasterFileCollection = args.input("tiles").unwrap();
        assert_eq!(coll.paths.len(), 2);
    }

    #[test]
    fn test_args_input_not_found() {
        let dir = setup_work_dir();
        let args = build_args_from(dir.path()).unwrap();
        let result: Result<File> = args.input("missing");
        assert!(matches!(result, Err(SpadeError::InputNotFound { .. })));
    }

    #[test]
    fn test_args_param_f64() {
        let dir = setup_work_dir();
        write_params(&dir, "resolution: 10.5\n");
        let args = build_args_from(dir.path()).unwrap();
        let val: f64 = args.param("resolution").unwrap();
        assert!((val - 10.5).abs() < f64::EPSILON);
    }

    #[test]
    fn test_args_param_string() {
        let dir = setup_work_dir();
        write_params(&dir, "method: bilinear\n");
        let args = build_args_from(dir.path()).unwrap();
        let val: String = args.param("method").unwrap();
        assert_eq!(val, "bilinear");
    }

    #[test]
    fn test_args_param_bool() {
        let dir = setup_work_dir();
        write_params(&dir, "normalize: true\n");
        let args = build_args_from(dir.path()).unwrap();
        let val: bool = args.param("normalize").unwrap();
        assert!(val);
    }

    #[test]
    fn test_args_param_not_found() {
        let dir = setup_work_dir();
        let args = build_args_from(dir.path()).unwrap();
        let result: std::result::Result<f64, _> = args.param("missing");
        assert!(matches!(result, Err(SpadeError::ParamNotFound { .. })));
    }

    #[test]
    fn test_args_has_input() {
        let dir = setup_work_dir();
        create_input_file(&dir, "source", "data.tif", b"test");
        let args = build_args_from(dir.path()).unwrap();
        assert!(args.has_input("source"));
        assert!(!args.has_input("missing"));
    }

    #[test]
    fn test_args_has_param() {
        let dir = setup_work_dir();
        write_params(&dir, "buffer: 30\n");
        let args = build_args_from(dir.path()).unwrap();
        assert!(args.has_param("buffer"));
        assert!(!args.has_param("missing"));
    }

    #[test]
    fn test_build_args_params_and_inputs() {
        let dir = setup_work_dir();
        write_params(&dir, "buffer: 30\n");
        create_input_file(&dir, "raster", "data.tif", b"test");
        let args = build_args_from(dir.path()).unwrap();
        assert!(args.has_param("buffer"));
        assert!(args.has_input("raster"));
    }

    #[test]
    fn test_args_input_directory() {
        // Directory type is constructed via FromInput directly.
        let dir = Directory::from_directory("/tmp/source".into()).unwrap();
        assert_eq!(dir.path, "/tmp/source");
    }
}
